### PR TITLE
fix #1343, support Storm emitDirect and DirectGrouping

### LINKED ISSUE #1343: support Storm OutputCollector.emitDirect and DirectGrouping

### PR DESCRIPTION
**Pull request by manuzhang**
remaining work

1. #1442 support Storm CustomStreamGrouping and CustomObject
2. full UTs
3. performance test
4. update documentation

all examples from storm-starter should work except `TridentWordCount` because of 1

### CODE REVIEW

**Comment by manuzhang** [comment]
Summary of this PR:

1. return the target `TaskId`s from output and allow user to output to a specific `TaskId.
2. expose subscriptions from `TaskActor` to user task through `TaskContext`  and add `getParallelism` to `Subscription`.
3. storm partitioners now also extends `StormPartitioner` which carries downstream component and stream information such that storm tasks could map a subscription to component.
4. #1373 happens because Storm uses `DirectGrouping` to send the number of emitted tuples between two `CoordinatedBolt`s for transaction besides the regular message flow. DirectGrouping is now ignored at graph build time and `emitDirect` is supported through the new output method from 1 and the mapping available from 2, 3.
5. storm has the use case where a processor only outputs to one of its subscribers (#1377) while gearpump processor broadcasts to all. This is resolved by filtering the input message at receiver side. The returned `TaskId`s are also filtered at sender side.

**Comment by huafengw** [comment]
why need to return the target TaskIds from output?

File: streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala [hunk 1]
```diff
@@ -69,14 +70,18 @@ class Subscription(
...
       }
     case mp: MulticastPartitioner =>
       (msg: Message) => {
         val partitions = mp.getPartitions(msg, parallelism, taskId.index)
-        partitions.map(partition => sendMessage(msg, partition)).sum
+        partitions.foldLeft((List.empty[TaskId], 0)) { (accum, partition) =>
+          val (tasks, msgsSent) = accum
+          val num = sendMessage(msg, partition)
+          (TaskId(processorId, partition) +: tasks) -> (msgsSent + num)
+        }
```
Comment (huafengw) [inline streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala]: In Subscription.scala, if ```life.contains(msg.timestamp)```, then the number of messages were sent is the length of ```partitions```, maybe you could avoid using foldLeft here.

File: streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala [hunk 2]
```diff
@@ -109,12 +105,25 @@ class TaskActor(
 
   def onStop() : Unit = task.onStop()
 
-  def output(msg : Message) : Unit = {
-    var count = 0
-    this.subscriptions.foreach{ subscription =>
-      count += subscription._2.sendMessage(msg)
+  def output(taskId: TaskId, msg: Message): Unit = {
+    subscriptions.find(_._1 == taskId.processorId) match {
```
Comment (huafengw) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: change subscriptions to Map?

File: streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala (hunk 1 above)
Comment (manuzhang) [inline streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala]: yes for now. My concern is that the return value of sendMessage could be changed to be more than 1 sometime.

**Comment by manuzhang** [comment]
@huafengw requirement from storm's CoordinatedBolt where messages are sent to a specific task

File: streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala (hunk 2 above)
Comment (manuzhang) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: I'm not sure. @clockfly any comments ?

**Comment by clockfly** [comment]
Can we avoid changing the Task interface?

**Comment by manuzhang** [comment]
@clockfly do you mean the existing `output` method ?

**Comment by clockfly** [comment]
I mean this one
https://github.com/gearpump/gearpump/pull/1441/files#diff-6732f2c78b22ca11a7680ec8153c1030

Task belongs to user interface, I am worried about the changes in class Task.

**Comment by manuzhang** [comment]
@clockfly no changes to `Task` interface now.  Please review again.

**Pull request closed without merging**